### ISSUE TITLE
Wrong code at -O2 on x86_64-linux_gnu since 4def99e

### ISSUE THREAD

**Issue by shao-hua-li** [llvm/llvm-project#70507]
clang at -O2 miscompiled the following code.

Bisected to 4def99e642806f3f8e94cb2aa9356d63cd233dd0, which was committed by @LebedevRI 

```console
% cat a.c
int printf(const char *, ...);
short a;
char b;
long c, d, e;
long *f;
static long **g = &f;
int h, i, j;
short(k)(short l, int m) { return m >= 2 || l > l << m; }
static long *n() {
  int o[3][36];
  h = 0;
  for (; h < 3; h++) {
    i = 0;
    for (; i < 4; i++) {
      j = 0;
      for (; j < 9; j++)
        o[h][i * j] = 55;
    }
  }
  b = 0;
  for (; b <= 2; b++) {
    c = 0;
    for (; c <= 2; c++)
      k(a++, o[2][6]) && (f = &e);
  }
  return &d;
}
int main() {
  n();
  **g = 0;
  printf("%d\n", 1);
}
%
% clang -O0 a.c && ./a.out
1
% clang -O2 a.c && ./a.out
%
```

**Comment by dtcxzyw** [comment]
Test case: https://godbolt.org/z/KMo396Kao
It seems like the SLP vectorizer handles poison propagation incorrectly.

**Comment by dtcxzyw** [comment]
Reduced test case: https://godbolt.org/z/YTaWjrqEE
Related commit: b9ad72ba056126d462d229877cdce1efa0761136

**Comment by dtcxzyw** [comment]
cc: @alexey-bataev

**Comment by alexey-bataev** [comment]
Fixed in 9da19e4340f21455b52d5768439cfbaca4112fe4

**Comment by dtcxzyw** [comment]
9da19e4 doesn't fully address this issue. Please take a look at https://godbolt.org/z/KMo396Kao.
```
  %16 = call i1 @llvm.vector.reduce.or.v8i1(<8 x i1> %15)
  %op.rdx = select i1 %16, i1 true, i1 %cmp2.i.2.2.i
  %op.rdx1 = select i1 %cmp.i.i, i1 true, i1 %cmp.i.i
  %op.rdx2 = select i1 %op.rdx, i1 true, i1 %op.rdx1
```
`%16` cannot be "head" of the logical-or chain because it is poison.

**Comment by alexey-bataev** [comment]
Must be fixed in 4c997e1536e4f5b43c38f6a4325441a7804d0da1

**Comment by dtcxzyw** [comment]
```
  %shl.i.2.2.i = shl nsw i32 %conv.i.2.2.i, %0
  %cmp2.i.2.2.i = icmp slt i32 %shl.i.2.2.i, %conv.i.2.2.i
  %16 = freeze <8 x i1> %15
  %17 = call i1 @llvm.vector.reduce.or.v8i1(<8 x i1> %16)
  %op.rdx = select i1 %17, i1 true, i1 %cmp2.i.2.2.i
  %op.rdx1 = select i1 %cmp.i.i, i1 true, i1 %cmp.i.i
  %op.rdx2 = select i1 %op.rdx, i1 true, i1 %op.rdx1
  br i1 %op.rdx2, label %18, label %for.cond15.preheader.i.n.exit_crit_edge
```
`%cmp2.i.2.2.i` will be simplified into `poison` in the follow-up `InstCombine` pass :(
I think we should also insert a freeze for `%cmp2.i.2.2.i`.

**Comment by alexey-bataev** [comment]
> ```
>   %shl.i.2.2.i = shl nsw i32 %conv.i.2.2.i, %0
>   %cmp2.i.2.2.i = icmp slt i32 %shl.i.2.2.i, %conv.i.2.2.i
>   %16 = freeze <8 x i1> %15
>   %17 = call i1 @llvm.vector.reduce.or.v8i1(<8 x i1> %16)
>   %op.rdx = select i1 %17, i1 true, i1 %cmp2.i.2.2.i
>   %op.rdx1 = select i1 %cmp.i.i, i1 true, i1 %cmp.i.i
>   %op.rdx2 = select i1 %op.rdx, i1 true, i1 %op.rdx1
>   br i1 %op.rdx2, label %18, label %for.cond15.preheader.i.n.exit_crit_edge
> ```
> 
> `%cmp2.i.2.2.i` will be simplified into `poison` in the follow-up `InstCombine` pass :( I think we should also insert a freeze for `%cmp2.i.2.2.i`.

Will try to check this tomorrow

**Comment by alexey-bataev** [comment]
> ```
>   %shl.i.2.2.i = shl nsw i32 %conv.i.2.2.i, %0
>   %cmp2.i.2.2.i = icmp slt i32 %shl.i.2.2.i, %conv.i.2.2.i
>   %16 = freeze <8 x i1> %15
>   %17 = call i1 @llvm.vector.reduce.or.v8i1(<8 x i1> %16)
>   %op.rdx = select i1 %17, i1 true, i1 %cmp2.i.2.2.i
>   %op.rdx1 = select i1 %cmp.i.i, i1 true, i1 %cmp.i.i
>   %op.rdx2 = select i1 %op.rdx, i1 true, i1 %op.rdx1
>   br i1 %op.rdx2, label %18, label %for.cond15.preheader.i.n.exit_crit_edge
> ```
> 
> `%cmp2.i.2.2.i` will be simplified into `poison` in the follow-up `InstCombine` pass :( I think we should also insert a freeze for `%cmp2.i.2.2.i`.

Seems like original code does the same. 
```
  %narrow.i.not.2.2.not.i = select i1 %cmp.i.i, i1 true, i1 %cmp2.i.2.2.i
  %11 = or i1 %narrow.i.not.2.2.not.i, %10
```

```
  %op.rdx = select i1 %17, i1 true, i1 %cmp2.i.2.2.i
  %op.rdx1 = select i1 %cmp.i.i, i1 true, i1 %cmp.i.i
  %op.rdx2 = select i1 %op.rdx, i1 true, i1 %op.rdx1
```
If `%cmp2.i.2.2.i` is poison and selected, `or` will propagate this poison value

**Comment by dtcxzyw** [comment]
> > ```
> >   %shl.i.2.2.i = shl nsw i32 %conv.i.2.2.i, %0
> >   %cmp2.i.2.2.i = icmp slt i32 %shl.i.2.2.i, %conv.i.2.2.i
> >   %16 = freeze <8 x i1> %15
> >   %17 = call i1 @llvm.vector.reduce.or.v8i1(<8 x i1> %16)
> >   %op.rdx = select i1 %17, i1 true, i1 %cmp2.i.2.2.i
> >   %op.rdx1 = select i1 %cmp.i.i, i1 true, i1 %cmp.i.i
> >   %op.rdx2 = select i1 %op.rdx, i1 true, i1 %op.rdx1
> >   br i1 %op.rdx2, label %18, label %for.cond15.preheader.i.n.exit_crit_edge
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > `%cmp2.i.2.2.i` will be simplified into `poison` in the follow-up `InstCombine` pass :( I think we should also insert a freeze for `%cmp2.i.2.2.i`.
> 
> Seems like original code does the same.
> 
> ```
>   %narrow.i.not.2.2.not.i = select i1 %cmp.i.i, i1 true, i1 %cmp2.i.2.2.i
>   %11 = or i1 %narrow.i.not.2.2.not.i, %10
> ```
> 
>  ```
>   %op.rdx = select i1 %17, i1 true, i1 %cmp2.i.2.2.i
>   %op.rdx1 = select i1 %cmp.i.i, i1 true, i1 %cmp.i.i
>   %op.rdx2 = select i1 %op.rdx, i1 true, i1 %op.rdx1
> ```
> 
> If `%cmp2.i.2.2.i` is poison and selected, `or` will propagate this poison value

The original code is well-defined because `%cmp.i.i` is true.
After SLP vectorization and instcombine, we got `%op.rdx2 = select i1 poison, i1 true, i1 true`, which is UB.
Therefore, there is still a miscompilation for SLP.

**Comment by alexey-bataev** [comment]
Ok, I see, will try to fix it today

**Comment by alexey-bataev** [comment]
> > > ```
> > >   %shl.i.2.2.i = shl nsw i32 %conv.i.2.2.i, %0
> > >   %cmp2.i.2.2.i = icmp slt i32 %shl.i.2.2.i, %conv.i.2.2.i
> > >   %16 = freeze <8 x i1> %15
> > >   %17 = call i1 @llvm.vector.reduce.or.v8i1(<8 x i1> %16)
> > >   %op.rdx = select i1 %17, i1 true, i1 %cmp2.i.2.2.i
> > >   %op.rdx1 = select i1 %cmp.i.i, i1 true, i1 %cmp.i.i
> > >   %op.rdx2 = select i1 %op.rdx, i1 true, i1 %op.rdx1
> > >   br i1 %op.rdx2, label %18, label %for.cond15.preheader.i.n.exit_crit_edge
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > `%cmp2.i.2.2.i` will be simplified into `poison` in the follow-up `InstCombine` pass :( I think we should also insert a freeze for `%cmp2.i.2.2.i`.
> > 
> > 
> > Seems like original code does the same.
> > ```
> >   %narrow.i.not.2.2.not.i = select i1 %cmp.i.i, i1 true, i1 %cmp2.i.2.2.i
> >   %11 = or i1 %narrow.i.not.2.2.not.i, %10
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ```
> >  %op.rdx = select i1 %17, i1 true, i1 %cmp2.i.2.2.i
> >  %op.rdx1 = select i1 %cmp.i.i, i1 true, i1 %cmp.i.i
> >  %op.rdx2 = select i1 %op.rdx, i1 true, i1 %op.rdx1
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > If `%cmp2.i.2.2.i` is poison and selected, `or` will propagate this poison value
> 
> The original code is well-defined because `%cmp.i.i` is true. After SLP vectorization and instcombine, we got `%op.rdx2 = select i1 poison, i1 true, i1 true`, which is UB. Therefore, there is still a miscompilation for SLP.

Fixed in 495ed8d8c8b3795dd51595aad7a192189f2cfeab

**Comment by dtcxzyw** [comment]
It works well now. Thanks for your fix! @alexey-bataev
